sdks/java/core/src/main/java/org/apache/beam/sdk/testing/TestStream.java

```diff
@@ -0,0 +1,223 @@
+   * {@link BoundedWindow#TIMESTAMP_MIN_VALUE}.
+   */
+  public static <T> Builder<T> create(Coder<T> coder) {
+    return new Builder<>(coder);
+  }
+
+  private TestStream(
+      Coder<T> coder,
+      List<TestSourceEvent<T>> events) {
+    this.coder = coder;
+    this.events = checkNotNull(events);
+  }
+
+  /**
+   * An incomplete {@link TestStream}. Elements added to this builder will be produced in
+   * sequence when the pipeline created by the {@link TestStream} is run.
+   */
+  public static class Builder<T> {
+    private final Coder<T> coder;
+    private final ImmutableList.Builder<TestSourceEvent<T>> events;
+    private Instant currentWatermark;
+
+    private Builder(Coder<T> coder) {
+      this.coder = coder;
+      events = ImmutableList.builder();
+
+      currentWatermark = BoundedWindow.TIMESTAMP_MIN_VALUE;
+    }
+
+    /**
+     * Adds the specified elements to the source with timestamp equal to the current watermark.
+     *
+     * @return this {@link TestStream.Builder}
+     */
+    @SafeVarargs
+    public final Builder<T> addElements(
+        T element, T... elements) {
+      TimestampedValue<T> firstElement = TimestampedValue.of(element, currentWatermark);
+      @SuppressWarnings("unchecked")
+      TimestampedValue<T>[] remainingElements = new TimestampedValue[elements.length];
+      for (int i = 0; i < elements.length; i++) {
+        remainingElements[i] = TimestampedValue.of(elements[i], currentWatermark);
+      }
+      return addElements(firstElement, remainingElements);
+    }
+
+    /**
+     * Adds the specified elements to the source with the provided timestamps.
+     *
+     * @return this {@link TestStream.Builder}
+     */
+    @SafeVarargs
+    public final Builder<T> addElements(
+        TimestampedValue<T> element, TimestampedValue<T>... elements) {
+      events.add(ElementEvent.add(element, elements));
+      return this;
+    }
+
+    /**
+     * Advance the watermark of this source to the specified instant.
+     *
+     * <p>The watermark must advance monotonically and to at most
+     * {@link BoundedWindow#TIMESTAMP_MAX_VALUE}.
+     *
+     * @return this {@link TestStream.Builder}
+     */
+    public Builder<T> advanceWatermarkTo(Instant newWatermark) {
+      checkArgument(newWatermark.isAfter(currentWatermark),
+          "The watermark must monotonically advance");
+      checkArgument(newWatermark.isBefore(BoundedWindow.TIMESTAMP_MAX_VALUE),
+          "The Watermark cannot progress beyond the maximum. Got: %s. Maximum: %s",
+          newWatermark,
+          BoundedWindow.TIMESTAMP_MAX_VALUE);
+      events.add(WatermarkEvent.<T>advanceTo(newWatermark));
+      currentWatermark = newWatermark;
+      return this;
+    }
+
+    /**
+     * Advance the processing time by the specified amount.
+     *
+     * @return this {@link TestStream.Builder}
+     */
+    public Builder<T> advanceProcessingTime(Duration amount) {
```

Comment:
Elements are output with an absolute timestamp, and the watermark is relative to these timestamps. However, Processing Time is always relative to whenever the pipeline happens to be executing, so users should not think of absolute processing times.